kolena/_experimental/dataset/_evaluation.py

```diff
@@ -216,27 +169,29 @@ def test(
     :param model: The name of the model to be used.
     :param results: Either a DataFrame or a list of tuples, where each tuple consists of
                     a eval configuration and a DataFrame.
-    :param on: The column(s) to merge on between datapoint DataFrame and result DataFrame
-
     :return None
     """
-    df_data = _fetch_dataset(dataset)
-    df_datapoints = _to_deserialized_dataframe(df_data, column=COL_DATAPOINT)
-    log.info(f"fetched {len(df_data)} for dataset {dataset}")
+    existing_dataset = load_dataset(dataset)
 
-    if isinstance(results, pd.DataFrame):
+    if isinstance(results, pd.DataFrame) or isinstance(results, Iterator):
         results = [(None, results)]
+    load_uuid = init_upload().uuid
 
     _validate_configs([cfg for cfg, _ in results])
-
-    all_results: List[Tuple[Optional[TYPE_EVALUATION_CONFIG], pd.DataFrame, pd.DataFrame]] = []
     for config, df_result_input in results:
         log.info(f"start evaluation with configuration {config}" if config else "start evaluation")
-        single_result = _align_datapoints_results(df_datapoints, df_result_input, on)
-        _validate_data(df_datapoints, single_result)
-        all_results.append((config, single_result, df_result_input))
-        log.info(f"completed evaluation with configuration {config}" if config else "completed evaluation")
-
-    df_results = _process_results(df_data, all_results, df_datapoints, on)
-    n_uploaded_results = _upload_results(model, df_results)
-    log.info(f"uploaded {n_uploaded_results} test results")
+        if isinstance(df_result_input, pd.DataFrame):
+            _validate_datapoint_id_columns(df_result_input, existing_dataset.id_fields)
+            df_results = _process_result(config, df_result_input, existing_dataset.id_fields)
+            upload_data_frame(df=df_results, batch_size=BatchSize.UPLOAD_RECORDS.value, load_uuid=load_uuid)
+        else:
+            id_column_validated = False
+            for df_result in df_result_input:
+                if not id_column_validated:
+                    _validate_datapoint_id_columns(df_result, existing_dataset.id_fields)
+                    id_column_validated = True
```

Comment:
Nit: theoretically, df_result in def_result_input could be of different shape

Reply:
the number of columns should be the same right?